src/pages/FrontPage.js

```diff
@@ -1,11 +1,214 @@
 import React from 'react';
 import CourseList from '../components/CourseList';
+import LessonFilter from '../components/filter/LessonFilter';
+
+const iconContext = require.context('lessonSrc/', true, /^\.\/[^\/]*\/logo-black\.png/);
+const lessonContext = require.context('onlyFrontmatter!lessonSrc/', true,
+  /^\.\/[^\/]*\/[^\/]*\/(?!README\.md$)[^\/]*\.md/);
 
 const FrontPage = React.createClass({
+  getInitialState() {
+    return {
+      filter: [],
+      tagGroups: [],
+      allCourses: [],
+      filteredCourses: []
+    };
+  },
+  componentWillMount() {
+    let allCourses = [];
+    let allTagGroups = [];
+    const paths = lessonContext.keys();
+
+    // Loop over all lessons. Create a list of all courses and a list of all tags.
+    paths.forEach(function (path) {
+      const courseName = path.slice(2, path.indexOf('/', 2));
+      const lessonFrontMatter = lessonContext(path).frontmatter;
+      const tagGroups = this.getTagGroupsFromLessonTags(lessonFrontMatter.tags);
+      const lesson = {name: lessonFrontMatter.title, tagGroups: tagGroups};
+
+      allTagGroups = allTagGroups.concat(tagGroups);
+      allCourses.push({
+        lessons: [lesson],
+        name: courseName,
+        iconPath: iconContext('./' + courseName + '/logo-black.png'),
+        path: path
+      });
+
+    }, this);
+
+    // Merge all tags that have same the name
+    // (e.g. platform: [not windows, only iPad] + platform: [not minecraft] =>
+    // platform: [not windows, only iPad, not minecraft])
+    const tagGroups = this.getListWithDistinctObjects(allTagGroups, this.mergeTagGroups);
+
+    // Merge all courses that have the same name by adding all lessons to the course
+    let courses = this.getListWithDistinctObjects(allCourses, function(existingCourse, newCourse){
+      existingCourse.lessons = existingCourse.lessons.concat(newCourse.lessons);
+    });
+
+    // Sort by number of lessons
+    courses = courses.sort(this.sortCourses);
+
+    // Update state
+    this.setState({
+      allCourses: courses,
+      filteredCourses: courses,
+      tagGroups: tagGroups
+    });
+  },
+  getTagGroupsFromLessonTags(lessonTagGroups) {
+    const tagGroups = [];
+    /* Tag structure:
+     lessonTagGroups = {
+      TagGroup1: [tagItem1, tagItem2, tagItem3],
+      TagGroup2: [tagItem4, tagItem5],
+      TagGroup3: [tagItem6]
+     } */
+    for (let groupName in lessonTagGroups) {
```

Comment:
Hvis du skal bruke `for ... in ...` på et object, så bør du ha med en sjekk for `hasOwnProperty`.
Et annet forslag er å skrive om og bruke en mer functional programming approach, f.eks.
```
const tagGroups = lessonTagGroups.reduce((result, groupName) => {
  ...
  result.push(.....);
  return result;
}, []);
```